---
title: "Supp_fig4"
author: "Christian Ayala Ortiz"
date: "`r Sys.Date()`"
output: html_document
---

# Load libraries

```{r setup, include=FALSE}
library(ggpubr)
library(tidyverse)
library(patchwork)

source('scripts/custom_functions.R')
```

# Load data

```{r}
mag_taxonomy <- load_mag_taxonomy()

metadata <- load_metadata()

mag_abundances <- load_mag_abundance(prev_min_samples = 0) %>% 
  decostand(method = 'tota', MARGIN = 1) %>% 
  rownames_to_column(var = 'asv_sample') %>% 
  pivot_longer(!asv_sample, names_to = 'bin', values_to = 'abundance') %>% 
  left_join(metadata) %>% 
  left_join(mag_taxonomy) %>% 
  group_by(month, asv_sample, Phylum) %>% 
  summarise(abundance = sum(abundance))


metaT_abundances <- read_csv('data/metaT/geTMM_normalized_matrix_2024.csv') %>% 
  rename_with(~str_remove(.x, '/xdisk.*results/')) %>% 
  pivot_longer(!Geneid, names_to = 'sample', values_to = 'tmm') %>% 
  mutate(sample = str_replace(sample, 'RNA', 'DNA'),
         bin = str_remove(Geneid, '_k.*')) %>% 
  left_join(metadata) %>% 
  left_join(mag_taxonomy)  %>% 
  group_by(month, asv_sample, Phylum) %>% 
  summarise(tmm = sum(tmm)) %>% 
  group_by(month, asv_sample) %>% 
  mutate(tmm_perc = tmm / sum(tmm))
```


# Plotting

```{r}
mag_abundance_plot <- mag_abundances %>% 
  ggplot() +
  geom_col(aes(x = asv_sample,
               y = abundance,
               fill = Phylum),
           color = 'black',
           linewidth = .1) +
  scale_fill_manual(values = get_palette('Paired', 18)) +
  scale_y_continuous(labels = scales::percent) +
  facet_grid(cols = vars(month),
             scales = 'free',
             space = 'free') +
  labs(y = 'Relative abundance\n(metagenomic reads)') +
  theme_bw() +
  custom_theme +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank())

mag_abundance_plot


metaT_abundance_plot <- metaT_abundances %>% 
  ggplot() +
  geom_col(aes(x = asv_sample,
               y = tmm_perc,
               fill = Phylum),
           color = 'black', linewidth = 0.1) +
  scale_fill_manual(values = get_palette('Paired', 18)) +
  scale_y_continuous(labels = scales::percent) +
  facet_grid(cols = vars(month),
             scales = 'free',
             space = 'free') +
  labs(y = 'Relative abundance\n(metatranscriptomics reads)') +
  theme_bw() +
  custom_theme +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank())

metaT_abundance_plot
```

# Final plot

```{r}
supp_fig5 <- mag_abundance_plot + metaT_abundance_plot +
  plot_layout(guides = 'collect', ncol = 1) +
  plot_annotation(tag_levels = 'A')

# supp_fig4

ggsave('output_figures/Supplementary_Figure_5.svg', supp_fig5, dpi = 300,
       height = 120, width = 180, units = 'mm')
ggsave('output_figures/Supplementary_Figure_5.png', supp_fig5, dpi = 300,
       height = 120, width = 180, units = 'mm')
```



